README


This archive contains the distribution OTRS-ForwardQueue,
version 0.12:

  Forwards the contents of an OTRS queue to a given email address.

This software is Copyright (c) 2014 by University of Manchester.

This is free software, licensed under:

  The GNU Affero General Public License, Version 3, November 2007


This README file was generated by Dist::Zilla::Plugin::Readme v5.020.



CHANGES
==================================================
Changes from 2013-10-02 00:00:00 +0000 to present.
==================================================

-----------------------------------------
version 0.12 at 2014-10-02 10:14:51 +0000
-----------------------------------------

  Change: f1c6002cd1aaadf2ebcdb2a88bc390ef51672ba8
  Author: Paul Waring <paul.waring@manchester.ac.uk>
  Date : 2014-10-02 11:14:36 +0000

    Bump version to 0.12 

  Change: 6147179ad90d4dd6cb61939e48239c33c441b648
  Author: Paul Waring <paul.waring@manchester.ac.uk>
  Date : 2014-10-02 11:13:59 +0000

    Update from address in OrigHeader 

  Change: 64d5c3e95307c5f1dd3c46810a9b296b2576a9f3
  Author: Paul Waring <paul.waring@manchester.ac.uk>
  Date : 2014-10-02 10:39:33 +0000

    Bump version to 0.11 

  Change: 84ab7ec7d9c6e0daba1c17f4787456d41bf4d2e9
  Author: Paul Waring <paul.waring@manchester.ac.uk>
  Date : 2014-10-02 10:39:16 +0000

    Configurable From address 

  Change: 4dc9ea5663932222d3499c99cd757f0182223149
  Author: Paul Waring <paul.waring@manchester.ac.uk>
  Date : 2014-10-01 09:55:45 +0000

    Bump version to 0.10 

  Change: 893b948a6204cfe8f1d3e2068cd23376d3722db7
  Author: Paul Waring <paul.waring@manchester.ac.uk>
  Date : 2014-10-01 09:55:31 +0000

    Fix incorrect method call 

  Change: 9b3501899c0db02f1eddfae0629c186ae43dbb55
  Author: Paul Waring <paul.waring@manchester.ac.uk>
  Date : 2014-09-30 12:54:29 +0000

    Bump to version 0.09 

  Change: 231266b0a3f1c2ee48f8e27f929cba57750dd24e
  Author: Paul Waring <paul.waring@manchester.ac.uk>
  Date : 2014-09-30 12:18:18 +0000

    Allow customer notification template to be changed 

  Change: 6ee14bb174fc80fd32805c87e2a3488638b75d35
  Author: Paul Waring <paul.waring@manchester.ac.uk>
  Date : 2014-09-11 13:34:32 +0000

    Update tested versions

    Have now tested with 3.3.9 

  Change: 91bad7d31c98c986dc1b22d2001df8ff1c4f625c
  Author: Paul Waring <paul@xk7.net>
  Date : 2014-04-29 21:34:14 +0000

    Bump version to 0.07 

  Change: cf25a6b975f380fcd6c889e0be160e439fa2d86c
  Author: Paul Waring <paul.waring@manchester.ac.uk>
  Date : 2014-04-24 10:48:40 +0000

    Make options read-only

    No need to change them after being supplied in constructor. 

  Change: 3f8a72ae22f80788602d26c03e41377be95d1bbe
  Author: Paul Waring <paul@xk7.net>
  Date : 2014-04-18 20:06:26 +0000

    Bump version to 0.06 

  Change: 74f2fbe971d99b4fd41b0b1d2bb2b0c896d5e774
  Author: Paul Waring <paul@xk7.net>
  Date : 2014-04-18 20:05:49 +0000

    Generate CHANGES file from Git log 

  Change: 6b89cd54bc401713833bf104f4e196075f0d6306
  Author: Paul Waring <paul@xk7.net>
  Date : 2014-04-18 17:07:41 +0000

    Bump version to 0.05 

  Change: 01f5a0fa1ddd1661e266bf4d0dea2a86877c04dc
  Author: Paul Waring <paul@xk7.net>
  Date : 2014-04-18 17:07:02 +0000

    Correct INI file syntax. 

  Change: 04c8c26c11346d9e438bf28e002a1f02d103cf6c
  Author: Paul Waring <paul@xk7.net>
  Date : 2014-04-18 17:03:56 +0000

    Set explict link for Bugtracker

    Failing to set link explicitly results in the default of rt.cpan.org
    being used.

    Uses Dist::Zilla::Plugin::Bugtracker to magically set the relevant
    metadata.

    Fixes #9 

  Change: b40062aea4b32af9450d068fc3168c984133276e
  Author: Paul Waring <paul@xk7.net>
  Date : 2014-04-18 17:03:30 +0000

    Update issues link to reflect change of repository name 

  Change: 775767cb59acf5206e94ad9d74040cf62d89fcd9
  Author: Paul Waring <paul@xk7.net>
  Date : 2014-04-18 13:29:37 +0000

    Bump version to 0.0.4 

  Change: f3eb93c924ff6021c4fb6d0a4543a345b9f90c12
  Author: Paul Waring <paul@xk7.net>
  Date : 2014-04-18 13:28:40 +0000

    Reduce README.md to bare minimum.

    All documentation is now contained within the module. README.md
    exists purely as an aid to those accessing the project through
    GitHub. 

  Change: f603e57221fb01ab6528190d6d188ac73cd7275f
  Author: Paul Waring <paul@xk7.net>
  Date : 2014-04-18 13:28:20 +0000

    Remove IO::Interactive from dependencies in README 

  Change: 0a5e59e49fbd53d9be98e049b03d309750c8cc3d
  Author: Paul Waring <paul@xk7.net>
  Date : 2014-04-18 13:25:50 +0000

    Convert documentation to use Pod::Weaver.

    Pod::Weaver automatically creates parts of the documentation using
    other parts of the code, e.g. VERSION is taken from $VERSION. 

  Change: 107f45f99e84d27d4432c07a04c47476d6842fbd
  Author: Paul Waring <paul@xk7.net>
  Date : 2014-04-18 13:17:51 +0000

    Bump version number 

  Change: 3a3825c516931554f5d7bd9591cd13ca97d5976d
  Author: Paul Waring <paul.waring@manchester.ac.uk>
  Date : 2014-04-17 11:37:27 +0000

    Update documentation

    Document process_queue method and remove IO::Interactive dependency
    (technically this applies to OTRS and not this module). 

  Change: d6c6ac473cf208dca96d1c9dfccf52c9e5bb5f21
  Author: Paul Waring <paul.waring@manchester.ac.uk>
  Date : 2014-04-16 15:35:38 +0000

    Bump version number 

  Change: 53137c123b777211e3702de98e5ac89fbeaade0b
  Author: Paul Waring <paul.waring@manchester.ac.uk>
  Date : 2014-04-15 12:11:36 +0000

    Update cron job documentation

    Don't need to specify local::lib on the command line, just the
    directory where it resides. 

  Change: 061c3334cc528fe9d37bd2e6909f808e4f4dffe7
  Author: Paul Waring <paul.waring@manchester.ac.uk>
  Date : 2014-04-14 16:30:04 +0000

    Document how to use module in a script under cron 

  Change: 7acb0b93fc3cd307214f1ac6b65456f7cbe934e1
  Author: Paul Waring <paul.waring@manchester.ac.uk>
  Date : 2014-04-14 15:54:43 +0000

    Remove environment file

    No longer used, simpler example is available in POD. 

  Change: 41e76717b20dffde39b9735377365f2aeb87d17a
  Author: Paul Waring <paul.waring@manchester.ac.uk>
  Date : 2014-04-14 11:41:17 +0000

    Documentation updates 

  Change: 9ce8ddc76e0a6f1605260ed6ab69875c4f180244
  Author: Paul Waring <paul.waring@manchester.ac.uk>
  Date : 2014-04-11 15:48:25 +0000


Makefile.PL

# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v5.020.
use strict;
use warnings;

use 5.014;

use ExtUtils::MakeMaker ;



my %WriteMakefileArgs = (
  "ABSTRACT" => "Forwards the contents of an OTRS queue to a given email address.",
  "AUTHOR" => "Paul Waring <paul.waring\@manchester.ac.uk>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "OTRS-ForwardQueue",
  "EXE_FILES" => [],
  "LICENSE" => "open_source",
  "NAME" => "OTRS::ForwardQueue",
  "PREREQ_PM" => {
    "Email::Sender::Simple" => 0,
    "Email::Sender::Transport::SMTP" => 0,
    "Email::Simple" => 0,
    "Email::Simple::Creator" => 0,
    "Moose" => 0,
    "Template" => 0,
    "namespace::autoclean" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "VERSION" => "0.12",
  "test" => {
    "TESTS" => ""
  }
);


my %FallbackPrereqs = (
  "Email::Sender::Simple" => 0,
  "Email::Sender::Transport::SMTP" => 0,
  "Email::Simple" => 0,
  "Email::Simple::Creator" => 0,
  "Moose" => 0,
  "Template" => 0,
  "namespace::autoclean" => 0,
  "strict" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);





lib/OTRS/ForwardQueue.pm
package OTRS::ForwardQueue;

use 5.014;

use strict;
use warnings;

use Moose;
use namespace::autoclean;

use Email::Sender::Simple;
use Email::Sender::Transport::SMTP;
use Email::Simple;
use Email::Simple::Creator;

use Template;

use Kernel::Config;
use Kernel::System::Encode;
use Kernel::System::Log;
use Kernel::System::Time;
use Kernel::System::Main;
use Kernel::System::DB;
use Kernel::System::Ticket;
use Kernel::System::Ticket::Article;

our $VERSION = '0.12';

has 'query' => (
  traits => ['Hash'],
  is => 'rw',
  isa => 'HashRef',
  required => 1,
  handles => {
    get_query => 'get',
    set_query => 'set',
  },
);

has 'options' => (
  traits => ['Hash'],
  is => 'ro',
  isa => 'HashRef',
  required => 1,
  handles => {
    get_option => 'get',
    exists_option => 'exists',
    defined_option => 'defined',
  },
);

sub process_queue
{
  my $self = shift;

  # Create all objects necessary for searching tickets
  # Taken from documentation for Kernel::System::Ticket
  my $ConfigObject = Kernel::Config->new();

  if ($self->exists_option('TempDir') && $self->defined_option('TempDir'))
  {
    $ConfigObject->Set( Key => 'TempDir', Value => $self->get_option('TempDir') );
  }

  my $EncodeObject = Kernel::System::Encode->new(
    ConfigObject => $ConfigObject,
  );

  my $LogObject = Kernel::System::Log->new(
    ConfigObject => $ConfigObject,
    EncodeObject => $EncodeObject,
  );

  my $TimeObject = Kernel::System::Time->new(
    ConfigObject => $ConfigObject,
    LogObject    => $LogObject,
  );

  my $MainObject = Kernel::System::Main->new(
    ConfigObject => $ConfigObject,
    EncodeObject => $EncodeObject,
    LogObject    => $LogObject,
  );

  my $DBObject = Kernel::System::DB->new(
    ConfigObject => $ConfigObject,
    EncodeObject => $EncodeObject,
    LogObject    => $LogObject,
    MainObject   => $MainObject,
  );

  my $TicketObject = Kernel::System::Ticket->new(
    ConfigObject       => $ConfigObject,
    LogObject          => $LogObject,
    DBObject           => $DBObject,
    MainObject         => $MainObject,
    TimeObject         => $TimeObject,
    EncodeObject       => $EncodeObject,
  );  
  
  # Always return results as an array, as we use Ticket ID to obtain any additional
  # information (results as a hash includes Ticket Number as well)
  $self->set_query('Result' => 'ARRAY');
  
  my @results = $TicketObject->TicketSearch(%{$self->query});
  
  foreach my $ticket_id (@results)
  {
    if ($self->exists_option('Debug') && $self->defined_option('Debug') && $self->get_option('Debug'))
    {
      print "Processing ticket ID: $ticket_id\n";
    }
    
    my %ticket = $TicketObject->TicketGet(
      TicketID => $ticket_id,
    );
    
    unless ($self->exists_option('DisableLocking') && $self->defined_option('DisableLocking') && $self->get_option('DisableLocking'))
    {
      # Lock ticket before proceeding, to prevent other users from accessing it
      my $lock_success = $TicketObject->TicketLockSet(
        Lock => 'lock',
        TicketID => $ticket_id,
        UserID => $self->get_query('UserID'),
        SendNoNotification => 1,
      );
    }
    
    unless ($self->exists_option('DisableEmail') && $self->defined_option('DisableEmail') && $self->get_option('DisableEmail'))
    {
      # First article in ticket will be the original user request - we need this for the
      # body of the forwarded email and the full From: field
      my %first_article = $TicketObject->ArticleFirstArticle(
        TicketID => $ticket_id,
      );
    
      my $from_address = $first_article{'From'};
      my $recipient = $self->get_option('ForwardTo');
      
      my $forward_email = Email::Simple->create(
        header => [
          To => $recipient,
          From => $from_address,
          Subject => $ticket{'Title'},
        ],
        body => $first_article{'Body'},
      );
      
      # Set additional mail options, including envelope from
      my %mail_options = (
        from => $first_article{'CustomerID'},
      );
      
      if ($self->exists_option('SMTP') && $self->defined_option('SMTP') && $self->get_option('SMTP'))
      {
        my $transport = Email::Sender::Transport::SMTP->new({
          host => $self->get_option('SMTPServer'),
        });
        
        $mail_options{'transport'} = $transport;
      }
      
      Email::Sender::Simple->send($forward_email, \%mail_options);
      
      if ($self->exists_option('NotifyCustomer') && $self->defined_option('NotifyCustomer') && $self->get_option('NotifyCustomer'))
      {
        # Produce the body of the response to the customer
        my $nc_tt = Template->new({
          INCLUDE_PATH => $self->get_option('TemplatesPath')
        }) || die "$Template::ERROR\n";
        
        my $nc_output = '';
        my $nc_vars = {
          ticket => \%ticket,
        };
				
				my $notify_template = 'notify_customer.tt';
				
				if ($self->exists_option('NotifyCustomerTemplate') && $self->defined_option('NotifyCustomerTemplate') && $self->get_option('NotifyCustomerTemplate'))
				{
					$notify_template = $self->get_option('NotifyCustomerTemplate');
				}
        
        $nc_tt->process($notify_template, $nc_vars, \$nc_output) || die $nc_tt->error() . "\n";
				
				my $from_address = $first_article{'ToRealname'};
				
				if ($self->exists_option('FromAddress') && $self->defined_option('FromAddress') && $self->get_option('FromAddress'))
				{
					$from_address = $self->get_option('FromAddress');
				}
        
        # Add a new article, which should be emailed automatically to the customer.
        # Remember that To/From are reversed here, since we are sending an email to
        # the customer who raised the ticket.
        my $article_id = $TicketObject->ArticleSend(
          TicketID => $ticket_id,
          ArticleType => 'email-external',
          SenderType => 'system',
          From => $from_address,
          To => $first_article{'From'},
          Subject => 'Ticket forwarded: ' . $ticket{'Title'},
          Body => $nc_output,
          Charset => 'ISO-8859-15',
          MimeType => 'text/plain',
          HistoryType => 'EmailCustomer',
          HistoryComment => 'Notified customer of ticket forwarding',
          UserID => $self->get_query('UserID'),
          AutoResponseType => 'auto reply',
          OrigHeader => {
            From => $from_address,
            To => $first_article{'From'},
            Subject => 'Ticket forwarded: ' . $ticket{'Title'},
          },
        );
      }
    }
        
    unless ($self->exists_option('DisableHistory') && $self->defined_option('DisableHistory') && $self->get_option('DisableHistory'))
    {
      # Log the change in the history
      my $history_success = $TicketObject->HistoryAdd(
        Name => $self->get_option('HistoryComment'),
        HistoryType => 'Misc',
        TicketID => $ticket_id,
        CreateUserID => $self->get_query('UserID'),
      );
    }
    
    unless ($self->exists_option('DisableClosing') && $self->defined_option('DisableClosing') && $self->get_option('DisableClosing'))
    {
      # Mark the ticket as successfully closed
      my $close_success = $TicketObject->TicketStateSet(
        State => 'closed successful',
        TicketID => $ticket_id,
        UserID => $self->get_query('UserID'),
        SendNoNotifications => 1,
      );
    }
  }
}

__PACKAGE__->meta->make_immutable;

1; # Magic true value required at end of module

=pod

=encoding UTF-8

=head1 NAME

OTRS::ForwardQueue - Forwards the contents of an OTRS queue to a given email address.

=head1 VERSION

version 0.12

=head1 SYNOPSIS

    use OTRS::ForwardQueue;

    %query = (
      Queues => ['MyQueue'],
      States => ['new', 'open'],
      Locks => ['unlock'],
      UserID => 1,
    );

    %options = (
      ForwardTo => 'nobody@example.org',
      TempDir => '/tmp',
      HistoryComment => 'Forward to other request system',
      SMTP => 1,
      SMTPServer => 'smtp.example.org',
      NotifyCustomer => 1,
      NotifyCustomerTemplate => 'notify_customer.tt',
      TemplatesPath => '/usr/local/templates',
      Debug => 1,
    );

    my $fq = OTRS::ForwardQueue->new('query' => \%query, 'options' => \%options);

    $fp->process_queue();

=head1 DESCRIPTION

This module queries the Open Technology Real Services (OTRS) ticket management
system for tickets matching the query provided and then forwards these
tickets to an email address, closing them in OTRS.

The following functions are provided:

=over

=item new(\%query, \%options)

Produced automatically by Moose, this is the constructor for the class.

=over

=item \%query

Reference to a hash which contains the query parameters. This takes the same
key/value pairs as the C<TicketSearch> function of C<Kernel::System::TicketSearch>,
except that the C<Result> value is always set to 'ARRAY'.

=item \%options

Required list of options which affect how the queue is processed.

=over

=item C<HistoryComment> (required): The comment left in the history of the ticket when it is forwarded to anther system.

=item C<TempDir> (optional): Override the temporary directory used by the OTRS cache. Probably needs to be set if you are not running the module as the web server user (e.g. apache). If you get errors about file permissions, try setting this to C<'/tmp'>.

=item C<ForwardTo> (required): The email address to forward tickets to.

=item C<DisableLocking> (optional): Set to 1 to disable locking of forwarded tickets. Default behaviour is to lock tickets.

=item C<DisableHistory> (optional): Set to 1 to disable leaving a comment in the history (effectively makes C<HistoryComment> redundant). Default behaviour is to add a comment.

=item C<DisableClosing> (optional): Set to 1 to disable closing the ticket after forwarding. Default behaviour is to mark ticket as 'closed successful'.

=item C<DisableEmail> (optional): Set to 1 to disable sending any emails, which effectively prevents forwarding of tickets. Default behaviour is to send email. Included as an option to allow module users to test their code before sending out emails.

=item C<SMTP> (optional): Set to 1 to use an SMTP server to send email, instead of the local MTA.

=item C<SMTPServer> (optional): Host name or IP address of the SMTP server to use. Only effective if C<SMTP> is set to 1.

=item C<NotifyCustomer> (optional): Set to 1 to create a new article on the ticket and notify the customer that it has been forwarded.

=item C<NotifyCustomerTemplate> (optional): Relative filename to template for customer notification. Required if C<NotifyCustomer> is set to 1.

=item C<TemplatesPath> (optional): Absolute path to template directory. Required if C<NotifyCustomerTemplate> is set.

=item C<Debug> (optional): Set to 1 to print extra debugging information, such as the IDs of forwarded tickets.

=back

=back

=item process_queue()

Processes the queue based on the options passed in the constructor.

=back

=head1 DEPENDENCIES

Perl version 5.14 or higher is required. You may be able to use the module with older versions of Perl, but this is neither tested nor supported.

This module requires the following modules:

=over 4

=item * L<Moose>

=item * L<namespace::autoclean>

=item * L<Email::Simple>

=item * L<Email::Sender>

=item * L<Template> - For dynamically producing the body of emails.

=back

Although some of the above modules are used for optional features, all the dependencies
must be installed as this module will attempt to import all of them.

You must also have the OTRS source installed and available via C<@INC>. This module has
been tested with OTRS 3.2.10 and 3.3.9.

=head1 RUNNING AS A CRON JOB

Running a script which uses this module as a cron job may require some additional tweaks.
The easiest way is to create a small wrapper script to set the various library paths
correctly, such as the one below:

    #!/bin/bash
    
    # Set this to the absolute path to your OTRS install, so those
    # modules can be loaded
    FQ_OTRS_LIB="-I/path/to/otrs"
    
    # Comment out this line if you are not using local::lib
    FQ_LOCAL_LIB="-I$HOME/perl5/lib/perl5"
    
    # Change this to the path of your script
    /usr/bin/perl "$FQ_OTRS_LIB" "$FQ_LOCAL_LIB" /path/to/script.pl

=head1 INCOMPATIBILITIES

None reported.

=head1 BUGS AND LIMITATIONS

No bugs have been reported.

Please report any bugs through the Github issue system:

L<https://github.com/pwaring/OTRS-ForwardQueue/issues>

=head1 AUTHOR

Paul Waring <paul.waring@manchester.ac.uk>

=head1 COPYRIGHT AND LICENSE

This software is Copyright (c) 2014 by University of Manchester.

This is free software, licensed under:

  The GNU Affero General Public License, Version 3, November 2007

=cut

__END__

# ABSTRACT: Forwards the contents of an OTRS queue to a given email address.

